File: python/demo/error_codes_demo.py
#!/usr/bin/env python3
"""
Python SDK - Error Codes Demo

This demo shows how error codes and source locations are included in validation errors.
"""

import json
import sys
import os

# Add the src directory to the path so we can import without installing
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))

from json_structure import SchemaValidator, InstanceValidator
from json_structure import (
    SCHEMA_NULL, SCHEMA_INVALID_TYPE, SCHEMA_REF_NOT_FOUND, SCHEMA_MIN_GREATER_THAN_MAX,
    INSTANCE_STRING_MIN_LENGTH, INSTANCE_REQUIRED_PROPERTY_MISSING, 
    INSTANCE_TYPE_MISMATCH, INSTANCE_STRING_PATTERN_MISMATCH
)

print("=== JSON Structure Python SDK - Error Codes Demo ===\n")

# Test 1: Schema validation errors with error codes and source locations
print("--- Schema Validation Errors ---\n")

# Use source text to enable source location tracking
schema_source = """{
    "$schema": "https://json-structure.org/meta/core/v0/#",
    "type": "object",
    "properties": {
        "count": {"type": "integer", "minimum": 10, "maximum": 5}
    },
    "required": ["missing_property"]
}"""

bad_schema = json.loads(schema_source)

print("Bad Schema:")
print(schema_source)
print()

schema_validator = SchemaValidator()
# Pass source_text to enable line/column tracking
schema_errors = schema_validator.validate(bad_schema, schema_source)

print(f"Valid: {len(schema_errors) == 0}\n")
print("Errors with error codes:")
print("-" * 80)

for error in schema_errors:
    print(f"  Error: {error.message}")
    print(f"         → Code:     {error.code}")
    print(f"         → Path:     {error.path}")
    if error.location and error.location.is_known:
        print(f"         → Location: Line {error.location.line}, Column {error.location.column}")
    print()

# Test 2: Instance validation errors with error codes
print("\n--- Instance Validation Errors ---\n")

schema = {
    "$schema": "https://json-structure.org/meta/core/v0/#",
    "type": "object",
    "properties": {
        "name": {"type": "string"},
        "age": {"type": "int32"}
    },
    "required": ["name", "age"]
}

instance = {
    "name": 123,        # Wrong type (should be string)
    # Missing required "age"
}

print("Schema:")
print(json.dumps(schema, indent=2))
print()
print("Instance:")
print(json.dumps(instance, indent=2))
print()

# Create instance validator with the schema in the constructor
instance_validator = InstanceValidator(schema)
# validate_instance returns a list of error strings
instance_errors = instance_validator.validate_instance(instance)

print(f"Valid: {len(instance_errors) == 0}\n")
print("Errors:")
print("-" * 80)

for error in instance_errors:
    # Instance errors are strings, not ValidationError objects
    print(f"  {error}")

print()

# Test 3: Show available error codes
print("\n--- Available Error Codes (sample) ---\n")
print("Schema Error Codes:")
print(f'  SCHEMA_NULL = "{SCHEMA_NULL}"')
print(f'  SCHEMA_INVALID_TYPE = "{SCHEMA_INVALID_TYPE}"')
print(f'  SCHEMA_REF_NOT_FOUND = "{SCHEMA_REF_NOT_FOUND}"')
print(f'  SCHEMA_MIN_GREATER_THAN_MAX = "{SCHEMA_MIN_GREATER_THAN_MAX}"')

print("\nInstance Error Codes:")
print(f'  INSTANCE_STRING_MIN_LENGTH = "{INSTANCE_STRING_MIN_LENGTH}"')
print(f'  INSTANCE_REQUIRED_PROPERTY_MISSING = "{INSTANCE_REQUIRED_PROPERTY_MISSING}"')
print(f'  INSTANCE_TYPE_MISMATCH = "{INSTANCE_TYPE_MISMATCH}"')
print(f'  INSTANCE_STRING_PATTERN_MISMATCH = "{INSTANCE_STRING_PATTERN_MISMATCH}"')

print("\n=== Demo Complete ===\n")

File: java/src/main/java/org/json_structure/demo/ErrorCodesDemo.java
package org.json_structure.demo;

import org.json_structure.validation.*;

/**
 * Java SDK - Error Codes Demo
 * 
 * This demo shows how error codes are included in validation errors.
 */
public class ErrorCodesDemo {
    public static void main(String[] args) {
        System.out.println("=== JSON Structure Java SDK - Error Codes Demo ===\n");

        // Test 1: Instance validation errors
        System.out.println("--- Instance Validation Errors ---\n");

        String schemaJson = """
            {
                "type": "object",
                "properties": {
                    "name": { "type": "string", "minLength": 3 },
                    "age": { "type": "int32", "minimum": 0 },
                    "email": { "type": "string", "pattern": "^[a-z]+@[a-z]+\\\\.[a-z]+$" }
                },
                "required": ["name", "age"]
            }
            """;

        String instanceJson = """
            {
                "name": "AB",
                "email": "INVALID"
            }
            """;

        System.out.println("Input JSON:");
        String[] lines = instanceJson.split("\n");
        for (int i = 0; i < lines.length; i++) {
            System.out.printf("  Line %d: %s%n", i + 1, lines[i]);
        }
        System.out.println();

        ValidationOptions options = new ValidationOptions();
        options.setStopOnFirstError(false);

        InstanceValidator instanceValidator = new InstanceValidator(options);
        ValidationResult result = instanceValidator.validate(instanceJson, schemaJson);

        System.out.printf("Valid: %s%n%n", result.isValid());
        System.out.println("Errors with error codes:");
        System.out.println("-".repeat(80));

        for (ValidationError error : result.getErrors()) {
            System.out.printf("  Error: %s%n", error.getMessage());
            System.out.printf("         → Code:     %s%n", error.getCode());
            System.out.printf("         → Path:     %s%n", error.getPath());
            if (error.getLocation() != null) {
                System.out.printf("         → Location: Line %d, Column %d%n", 
                    error.getLocation().getLine(), error.getLocation().getColumn());
            }
            System.out.println();
        }

        // Test 2: Schema validation errors
        System.out.println("\n--- Schema Validation Errors ---\n");

        String badSchema = """
            {
                "type": "object",
                "properties": {
                    "count": { "type": "integer", "minimum": 10, "maximum": 5 }
                },
                "required": ["missing_property"]
            }
            """;

        System.out.println("Bad Schema:");
        String[] schemaLines = badSchema.split("\n");
        for (int i = 0; i < schemaLines.length; i++) {
            System.out.printf("  Line %d: %s%n", i + 1, schemaLines[i]);
        }
        System.out.println();

        SchemaValidator schemaValidator = new SchemaValidator(options);
        ValidationResult schemaResult = schemaValidator.validate(badSchema);

        System.out.printf("Valid: %s%n%n", schemaResult.isValid());
        System.out.println("Errors with error codes:");
        System.out.println("-".repeat(80));

        for (ValidationError error : schemaResult.getErrors()) {
            System.out.printf("  Error: %s%n", error.getMessage());
            System.out.printf("         → Code:     %s%n", error.getCode());
            System.out.printf("         → Path:     %s%n", error.getPath());
            if (error.getLocation() != null) {
                System.out.printf("         → Location: Line %d, Column %d%n", 
                    error.getLocation().getLine(), error.getLocation().getColumn());
            }
            System.out.println();
        }

        // Test 3: Show available error codes
        System.out.println("\n--- Available Error Codes (sample) ---\n");
        System.out.println("Schema Error Codes:");
        System.out.printf("  SCHEMA_NULL = \"%s\"%n", ErrorCodes.SCHEMA_NULL);
        System.out.printf("  SCHEMA_INVALID_TYPE = \"%s\"%n", ErrorCodes.SCHEMA_INVALID_TYPE);
        System.out.printf("  SCHEMA_REF_NOT_FOUND = \"%s\"%n", ErrorCodes.SCHEMA_REF_NOT_FOUND);
        System.out.printf("  SCHEMA_MIN_GREATER_THAN_MAX = \"%s\"%n", ErrorCodes.SCHEMA_MIN_GREATER_THAN_MAX);

        System.out.println("\nInstance Error Codes:");
        System.out.printf("  INSTANCE_STRING_MIN_LENGTH = \"%s\"%n", ErrorCodes.INSTANCE_STRING_MIN_LENGTH);
        System.out.printf("  INSTANCE_REQUIRED_PROPERTY_MISSING = \"%s\"%n", ErrorCodes.INSTANCE_REQUIRED_PROPERTY_MISSING);
        System.out.printf("  INSTANCE_TYPE_MISMATCH = \"%s\"%n", ErrorCodes.INSTANCE_TYPE_MISMATCH);
        System.out.printf("  INSTANCE_STRING_PATTERN_MISMATCH = \"%s\"%n", ErrorCodes.INSTANCE_STRING_PATTERN_MISMATCH);
    }
}
